src/libraries/System.Text.RegularExpressions/tests/UnitTests/SymbolicRegexTests.cs

```diff
@@ -79,7 +80,16 @@ public static IEnumerable<object[]> SafeThresholdTests_MemberData()
                 // lower bound int.MaxValue is never unfolded and treated as infinity
                 ("(a{2147483647,})", 2),
                 // typical case that blows up the DFA size to 2^100 when .* is added at the beginnig (below)
-                ("a.{100}b", 103)
+                ("a.{100}b", 103),
+                // these patterns are regression tests to make sure that right concatenations are propagated over alternations
+                ("a.{10}|bb", 14),
+                ("(a.{10}|bb)c", 15),
+                ("((a.{10}|bb)c)d", 16),
+                ("(((a.{10}|bb)c)d)*", 16),
+                // involves a deadend in form of an empty character class [\d-[\w]] that contributes with no states to an NFA
+                (@"([\d-[\w]]{10}|bc)", 3),
+                // in contrast this pattern has no deadends because [\w-[\d]] is nonempty
+                (@"([\w-[\d]]{10}|bc)", 13)
```

Comment:
Beyond these unit tests validating safe size, are we missing functional tests, i.e. patterns/inputs that would be producing the wrong answer on NonBacktracking before this PR?

Reply:
I think the `StressTestNfaMode` test would not have entered NFA mode on some of the new patterns similar to `.*?((a.{10}|bb)c)d` but for larger bounds like `.*?((a.{30}|bb)c)d` as the NFA mode relies on `SymbolicRegexNode.EnumerateAlternationBranches`. I will investigate that, and add some patterns into `StressTestNfaMode_TestData`.

In general it would be useful to have an AppContext variable, say,  REGEX_NONBACKTRACKING_NFA_THRESHOLD, to lower the bound
SymbolicRegexThresholds.NfaThreshold temporarily in those tests, to make sure NFA mode is triggered, but whether it is actually triggered or not is only visible from the public view, by termination, not actually sure how to test for NFA mode being entered there.

Overall, I think the NFA mode needs better testing.

Reply:
I mean, it would have entered the mode itself but not actually produced the NFA transitions it is supposed to in that mode, thus potentially causing nontermination and not being safe.

Reply:
Currently the `StressTestNfaMode` does not seem to trigger the difficult cases that would expose the bug, I'm thinking that the _random input generation_ there is too simplistic. I'll investigate this more today with @olsaarik.

Reply:
Thanks.

Reply:
I was going to ask the same question about whether there's a way to allow tests to run against the NFA mode. My understanding is that most functional tests won't exercise it today. If it's only for testing, it wouldn't be an app context flag- those are for public use. It would likely be simplest to just use reflection in the tests to modify a threshold. Then all the functional tests that run against non backtracking would run against both regular and forced NFA. Would that be helpful?

Reply:
I have been trying to come up with a functional test that exposes the above "bug" - or so I thought. _So far I have not been able to_. The main reason is that the backtracking-semantics in the first phase initially actually propagates right conjunctions over left disjunctions through **pruning** `PruneLowerPriorityThanNullability` (while the unit test is more simplistic and runs the exploration in nonbacktracking-semantics mode that gives more states in general). But in the **reverse direction** in phase 2 the problem should still come up, but even there I've not been able to come up with the right kind of regex. e.g., I've tried with regexes like 
`^.{20}(?:a|AA|DCB).*` that correspond to the hard case `(?:a|AA|BCD).{20}$` (but in reverse in the 2nd phase).

Many simplifications kick in that essentially avoid the problem. Still working on it, yesterday at some point I almost thought to **close** the PR before I understand the situation better and revisit this later.

Regarding the 

_it wouldn't be an app context flag- those are for public use_

I was thinking that it might be useful to have such a public flag to allow this limit to be configurable in general, not just for testing, but I'm not sure actually how useful that would be.

Reply:
Let me add a potential legitimate reason for such a flag, that @olsaarik pointed out in a meeting we had yesterday. When someone runs the matcher on a machine with a lot of memory to spare but in a time-critical situation and runs into the `10_000` limit just before the DFA is almost done but the NFA mode kicks in, eg. for a regex like `"a.{13)b"` where the DFA has roughly `2^14` states and the limit could be `17_000` in this case, allowing the limit to be configurable would make sense.